File: polynomial/polynomial.cpp
#include "polynomial.h"

#include <sstream>
#include <iomanip>
#include <cmath>
using std::stringstream;
using std::showpos;
using std::fixed;
using std::setprecision;

// Parameter: a size_t degree
// Constructor: Creates a polynomial of given degree, initializing all coefficients to 0
Polynomial::Polynomial(size_t degree) : _degree(degree), _coefficients(new float[_degree + 1]) {
    for(size_t i = 0; i < _degree + 1; i++) {
        _coefficients[i] = 0.0;
    }
}

// Parameters: a size_t degree, float* coefficient
// Constructor: Creates a polynomial of given degree with the provided coefficient values
Polynomial::Polynomial(size_t degree, const float* coefficients) : _degree(degree), _coefficients(new float[_degree + 1]) {
    for (size_t i = 0; i < _degree + 1; i++) {
        _coefficients[i] = coefficients[i];
    }
}

// Parameter: an other polynomial
// Copy constructor: Creates a deep copy of another polynomial
Polynomial::Polynomial(const Polynomial& other) : _degree(other._degree), _coefficients(new float[other._degree + 1]) {
    for (size_t i = 0; i < _degree + 1; i++) {
        _coefficients[i] = other._coefficients[i];
    }
}

// Destructor: Cleans up dynamically allocated memory
Polynomial::~Polynomial() {
    _degree = 0;
    delete[] _coefficients;
}

// Parameter: a rhs polynomial
// Copy assignment operator: Assigns values from another polynomial, performing deep copy
const Polynomial& Polynomial::operator=(const Polynomial& rhs) {
    delete[] _coefficients;
    _degree = rhs._degree;
    _coefficients = new float[_degree + 1];
    for(size_t i = 0; i <= _degree; i++) {
        _coefficients[i] = rhs._coefficients[i];
    }
    return *this;
}

// Parameter: a rhs polynomial
// Adds two polynomials term by term. Supports different degrees by treating missing terms as 0
const Polynomial Polynomial::Sum(const Polynomial& rhs)const {
    size_t bigDegree = std::max(_degree, rhs._degree);                   // finds the polynomial that has more coefficients
    float* values = new float[bigDegree + 1];
    for (size_t i = 0; i <= bigDegree; i++) {
        float lhsCoeff = 0.0f;
        float rhsCoeff = 0.0f;

        if (i<= _degree) {                                                   // checks if the current index in within bounds of the current polynomial
            lhsCoeff = _coefficients[i];
        }
        if (i <= rhs._degree) {
            rhsCoeff = rhs._coefficients[i];
        }
        values[i] = lhsCoeff + rhsCoeff;
    }

    Polynomial result(bigDegree, values);
    delete[] values;
    return result;
}

// Parameter: a rhs polynomial
// Subtracts rhs polynomial from this polynomial term by term. Like addition, handles different degrees
const Polynomial Polynomial::Subtract(const Polynomial& rhs)const {
    size_t bigDegree = std::max(_degree, rhs._degree);                   // finds the polynomial that has more coefficients
    float* values = new float[bigDegree + 1];
    for (size_t i = 0; i <= bigDegree; i++) {
        float lhsCoeff = 0.0f;
        float rhsCoeff = 0.0f;

        if (i <= _degree) {                                                  // checks if the current index in within bounds of the current polynomial
            lhsCoeff = _coefficients[i];
        }
        if (i <= rhs._degree) {
            rhsCoeff = rhs._coefficients[i];
        }
        values[i] = lhsCoeff - rhsCoeff;
    }

    Polynomial result(bigDegree, values);
    delete[] values;
    return result;
}

// Returns the additive inverse of the polynomial (i.e., multiplies every coefficient by -1)
const Polynomial Polynomial::Minus()const {
    float* inverse = new float[_degree + 1];
    for(size_t i = 0; i <= _degree; i++) {
        inverse[i] = -_coefficients[i];
    }
    Polynomial result(_degree, inverse);
    delete[] inverse;
    return result;
}

// Parameter: a rhs polynomial
// Multiplies two polynomials using the distributive property (nested loop over terms)
// This implements classical polynomial multiplication (FOIL method)
const Polynomial Polynomial::Multiply(const Polynomial& rhs)const {
    size_t degrees = _degree + rhs._degree;
    float* values = new float[degrees + 1]();      // initialize to zero
    for(size_t i = 0; i < _degree + 1; i++) {
        for(size_t j = 0; j < rhs._degree + 1; j++) {
            values[i + j] += _coefficients[i] * rhs._coefficients[j];
        }
    }
    Polynomial result(degrees, values);
    delete[] values;
    return result;
}

// Parameter: a rhs polynomial
// Performs polynomial long division: this / rhs
// Returns the quotient polynomial and ignores the remainder
// Implements the long division algorithm: subtract scaled divisor from dividend step-by-step
const Polynomial Polynomial::Divide(const Polynomial& rhs)const{
    size_t degreeDiff = _degree - rhs._degree;
    float* quoCoeff = new float[degreeDiff + 1]();
    Polynomial remainder = *this;
    for (int i = degreeDiff; i >= 0; i--) {
        float coeff = remainder._coefficients[i + rhs._degree] / rhs._coefficients[rhs._degree];  // divides heading term of the remainder by the leading term of the divisor
        quoCoeff[i] = coeff;

        for (size_t j = 0; j <= rhs._degree; j++){
            remainder._coefficients[i + j] -= coeff * rhs._coefficients[j];                       //Subtract rhs multiplied by the current coefficient from the remainder
        }
    }

    Polynomial quotient(degreeDiff, quoCoeff);
    delete[] quoCoeff;
    return quotient;
}

// Parameters: a rhs polynomial, a remainder polynomial
// Return: a polynomial that will contain the division of this with rhs
// The remainder will be placed in parameter by reference remainder
const Polynomial Polynomial::Divide(const Polynomial& rhs, Polynomial& remainder)const{
    int degreeDiff = _degree - rhs._degree;
    if (degreeDiff < 0) {                                                                           // If the degree of the divisor is greater, quotient is 0
        remainder = *this;
        return Polynomial(0);
    }

    float* quoCoeff = new float[degreeDiff + 1]();
    remainder = *this;
    for (int i = degreeDiff; i >= 0; i--) {
        float coeff = remainder._coefficients[i + rhs._degree] / rhs._coefficients[rhs._degree];    // divides heading term of the remainder by the leading term of the divisor
        quoCoeff[i] = coeff;

        for (size_t j = 0; j <= rhs._degree; j++) {
            remainder._coefficients[i + j] -= coeff * rhs._coefficients[j];                        //Subtract rhs multiplied by the current coefficient from the remainder
        }
    }

    Polynomial quotient(degreeDiff, quoCoeff);
    delete[] quoCoeff;
    return quotient;
}

// Computes the derivative of the polynomial using the power rule:
// For each term a_{i+1} * x^{i+1}, the derivative is (i+1) * a_{i+1} * x^i
// This loop builds the resulting polynomial: P'(x) = ∑ (i+1) * a_{i+1} * x^i
const Polynomial Polynomial:: Derive()const {
    float* derivative = new float[_degree];
    for(size_t i = 0; i < _degree ; i++) {
        derivative[i] = _coefficients[i + 1] * (i + 1);
    }
    Polynomial result(_degree - 1, derivative);
    delete[] derivative;
    return result;
}

// Parameter: a float x
// Evaluates the polynomial at a given x using direct computation
float Polynomial::Evaluate(float x)const {
    float result = 0.0;
    for(size_t i = 0; i < _degree + 1; i++) {
        result += _coefficients[i] * pow(x,i);            // Adds a_i * x^i to the result for each term
    }
    return result;
}

// Parameters: a float start, a float end
// Computes the definite integral of the polynomial from 'start' to 'end' using the power rule of integration:
// ∫ a*x^n dx = (a/(n+1)) * x^(n+1), then evaluate the result at 'end' and 'start' and subtract
float Polynomial::Integrate(float start, float end)const{
    float result = 0.0f;
    for(size_t i = 0; i <= _degree; i++ ) {
        float exponent = static_cast<float>(i + 1);
        result += _coefficients[i] * (pow(end, exponent) - pow(start, exponent)) / exponent;          // For each term a*x^n, compute a * [(end^(n+1) - start^(n+1)) / (n+1)]
    }
    return result;
}

// Parameter: a rhs polynomial
// Compares this polynomial to another for equality by checking each coefficient.
bool Polynomial:: Equals(const Polynomial& rhs)const {
    for(size_t i = 0; i < _degree + 1; i++) {
        if(_coefficients[i] != rhs._coefficients[i]) return false;
    }
    return true;
}

// GIVEN. Return: a string representation of the polynomial.
string Polynomial::ToString()const{
    stringstream ss;
    for (size_t i = _degree; i > 0; i--) {
        ss << showpos << fixed << setprecision(2) << _coefficients[i] << "x^" << i << " ";
    }
    ss << showpos << fixed << setprecision(2) << _coefficients[0];
    return ss.str();
}

// Parameter: an output stream to write to
// GIVEN. Writes a polynomial from the output stream.
ostream& Polynomial::Write(ostream& output)const{
    output << _degree << " ";
    for (size_t i = 0; i < _degree + 1; i++) {
        output << _coefficients[i] << " ";
    }
    return output;
}

// Parameter: an input stream to read from
// GIVEN. Reads a polynomial from the input stream.
istream& Polynomial::Read(istream& input){
    size_t degree;
    input >> degree;
    if (input.fail()){
        return input;
    }
    float* coefficients = new float[degree + 1];
    for (size_t i = 0; i < degree + 1; i++) {
        input >> coefficients[i];
        if (input.fail()){
            delete[] coefficients;
            return input;
        }
    }

    if (degree != _degree){
        if (_coefficients){
            delete[] _coefficients;
        }
        _degree = degree;
        _coefficients = coefficients;
    }else{
        for (size_t i = 0; i < _degree + 1; i++) {
            _coefficients[i] = coefficients[i];
        }
        delete[] coefficients;
    }
    return input;
}

File: polynomial/polynomial.py
from typing import List, Tuple
import math


class Polynomial:
    """
    A class representing a polynomial with arithmetic operations.
    """

    def __init__(self, degree: int, coefficients: List[float] = None):
        """
        Initialize a polynomial with given degree and optional list of coefficients.

        Parameters:
            degree (int): The degree of the polynomial.
            coefficients (List[float], optional): Coefficients in ascending order of power (index 0 = constant).
                                                  If not provided, initializes to 0 for all terms.
        """
        self._degree = degree
        if coefficients is None:
            self._coefficients = [0.0] * (degree + 1)
        else:
            self._coefficients = coefficients[:degree + 1]

    def copy(self) -> 'Polynomial':
        """
        Create a deep copy of the polynomial.
        Returns: A new polynomial with the same coefficients and degree.
        """
        return Polynomial(self._degree, self._coefficients.copy())

    def __eq__(self, other: 'Polynomial') -> bool:
        """
        Check equality between two polynomials.
        Parameters: The polynomial to compare with.
        Returns: True if coefficients are the same, False otherwise.
        """
        return self._coefficients == other._coefficients

    def sum(self, rhs: 'Polynomial') -> 'Polynomial':
        """
        Add two polynomials.
        Parameters: The right-hand side polynomial to add.
        Returns: The result of the addition.
        """
        big_degree = max(self._degree, rhs._degree)
        result_coeffs = [0.0] * (big_degree + 1)

        for i in range(big_degree + 1):
            lhs_val = self._coefficients[i] if i <= self._degree else 0.0
            rhs_val = rhs._coefficients[i] if i <= rhs._degree else 0.0
            result_coeffs[i] = lhs_val + rhs_val

        return Polynomial(big_degree, result_coeffs)

    def subtract(self, rhs: 'Polynomial') -> 'Polynomial':
        """
        Subtract another polynomial from this one.
        Parameters: The polynomial to subtract (this - rhs).
        Returns: The result of the subtraction.
        """
        big_degree = max(self._degree, rhs._degree)
        result_coeffs = [0.0] * (big_degree + 1)

        for i in range(big_degree + 1):
            lhs_val = self._coefficients[i] if i <= self._degree else 0.0
            rhs_val = rhs._coefficients[i] if i <= rhs._degree else 0.0
            result_coeffs[i] = lhs_val - rhs_val

        return Polynomial(big_degree, result_coeffs)

    def minus(self) -> 'Polynomial':
        """
        Return the negation of the polynomial.
        Returns: The negated polynomial (-1 * this).
        """
        result_coeffs = [-c for c in self._coefficients]
        return Polynomial(self._degree, result_coeffs)

    def multiply(self, rhs: 'Polynomial') -> 'Polynomial':
        """
        Multiply two polynomials.
        Parameters: The right-hand side polynomial to multiply.
        Returns: The product polynomial.
        """
        new_degree = self._degree + rhs._degree
        result_coeffs = [0.0] * (new_degree + 1)

        for i in range(self._degree + 1):
            for j in range(rhs._degree + 1):
                result_coeffs[i + j] += self._coefficients[i] * rhs._coefficients[j]

        return Polynomial(new_degree, result_coeffs)

    def divide(self, rhs: 'Polynomial') -> 'Polynomial':
        """
        Perform polynomial division and return only the quotient.
        Parameters: The divisor polynomial.
        Returns: The quotient polynomial.
        """
        quotient, _ = self._divide_with_remainder(rhs)
        return quotient

    def divide_with_remainder(self, rhs: 'Polynomial') -> Tuple['Polynomial', 'Polynomial']:
        """
        Divide the polynomial and return both the quotient and remainder.
        Parameters: The divisor polynomial.
        Returns: Tuple[Polynomial, Polynomial]: (quotient, remainder)
        """
        return self._divide_with_remainder(rhs)

    def _divide_with_remainder(self, rhs: 'Polynomial') -> Tuple['Polynomial', 'Polynomial']:
        """
        Internal method to divide polynomials using long division.
        Parameters: The divisor polynomial.
        Returns: Tuple[Polynomial, Polynomial]: (quotient, remainder)
        """
        if rhs._degree > self._degree:
            return Polynomial(0, [0.0]), self.copy()

        degree_diff = self._degree - rhs._degree
        quo_coeffs = [0.0] * (degree_diff + 1)
        remainder = self._coefficients[:]

        for i in reversed(range(degree_diff + 1)):
            coeff = remainder[i + rhs._degree] / rhs._coefficients[rhs._degree]
            quo_coeffs[i] = coeff
            for j in range(rhs._degree + 1):
                remainder[i + j] -= coeff * rhs._coefficients[j]

        return Polynomial(degree_diff, quo_coeffs), Polynomial(len(remainder) - 1, remainder)

    def derive(self) -> 'Polynomial':
        """
        Compute the derivative of the polynomial.
        Returns: The first derivative of the polynomial.
        """
        if self._degree == 0:
            return Polynomial(0, [0.0])

        derived_coeffs = [(i + 1) * self._coefficients[i + 1] for i in range(self._degree)]
        return Polynomial(self._degree - 1, derived_coeffs)

    def evaluate(self, x: float) -> float:
        """
        Evaluate the polynomial at a specific value of x.
        Parameters: The input value.
        Returns: The result of the polynomial evaluated at x.
        """
        return sum(self._coefficients[i] * math.pow(x, i) for i in range(self._degree + 1))

    def integrate(self, start: float, end: float) -> float:
        """
        Compute the definite integral from start to end.
        Parameters:
            start (float): Lower limit of integration.
            end (float): Upper limit of integration.
        Returns: The definite integral result.
        """
        result = 0.0
        for i in range(self._degree + 1):
            exponent = i + 1
            result += self._coefficients[i] * (math.pow(end, exponent) - math.pow(start, exponent)) / exponent
        return result

    def __str__(self) -> str:
        """
        Convert the polynomial to a human-readable string.
        Returns: Polynomial as a string in descending order.
        """
        terms = []
        for i in reversed(range(1, self._degree + 1)):
            coeff = self._coefficients[i]
            terms.append(f"{coeff:+.2f}x^{i}")
        terms.append(f"{self._coefficients[0]:+.2f}")
        return ' '.join(terms)

    def read(self, input_string: str) -> bool:
        """
        Read polynomial data from a string of space-separated numbers.
        Parameters: The input string. First number is degree, followed by coefficients.
        Returns: True if successful, False if format is invalid.
        """
        try:
            parts = list(map(float, input_string.strip().split()))
            degree = int(parts[0])
            if len(parts[1:]) != degree + 1:
                return False
            self._degree = degree
            self._coefficients = parts[1:]
            return True
        except:
            return False

    def write(self) -> str:
        """
        Write polynomial data as a string.
        Returns: A string starting with degree, followed by coefficients.
        """
        return f"{self._degree} " + ' '.join(f"{c:.2f}" for c in self._coefficients)